src/Servers/Kestrel/Core/src/Internal/Http/ForContentLengthMessageBody.cs

```diff
@@ -0,0 +1,302 @@
+                return _readResult;
+            }
+
+            TryStart();
+
+            while (true)
+            {
+                // This isn't great. The issue is that TryRead can get a canceled read result
+                // which is unknown to StartTimingReadAsync. 
+                if (_context.RequestTimedOut)
+                {
+                    BadHttpRequestException.Throw(RequestRejectionReason.RequestBodyTimeout);
+                }
+
+                _readResult = await StartTimingReadAsync(cancellationToken);
+
+                if (_context.RequestTimedOut)
+                {
+                    Debug.Assert(_readResult.IsCanceled);
+                    BadHttpRequestException.Throw(RequestRejectionReason.RequestBodyTimeout);
+                }
+
+                if (_readResult.IsCanceled)
+                {
+                    if (Interlocked.CompareExchange(ref _userCanceled, 0, 1) == 1)
+                    {
+                        // Ignore the readResult if it wasn't by the user.
+                        break;
+                    }
+                }
+
+                var readableBuffer = _readResult.Buffer;
+                var readableBufferLength = readableBuffer.Length;
+                StopTimingRead(readableBufferLength);
+
+                if (_readResult.IsCompleted)
+                {
+                    // OnInputOrOutputCompleted() is an idempotent method that closes the connection. Sometimes
+                    // input completion is observed here before the Input.OnWriterCompleted() callback is fired,
+                    // so we call OnInputOrOutputCompleted() now to prevent a race in our tests where a 400
+                    // response is written after observing the unexpected end of request content instead of just
+                    // closing the connection without a response as expected.
+                    _context.OnInputOrOutputCompleted();
+
+                    BadHttpRequestException.Throw(RequestRejectionReason.UnexpectedEndOfRequestContent);
+                }
+
+                if (readableBufferLength > 0)
+                {
+                    break;
+                }
+            }
+
+            // handle cases where we send more data than the content length
+            if (_readResult.Buffer.Length > _inputLength)
+            {
+                _readResult = new ReadResult(_readResult.Buffer.Slice(0, _inputLength), _readResult.IsCanceled, isCompleted: true);
+
+            }
+            else if (_readResult.Buffer.Length == _inputLength)
+            {
+                _readResult = new ReadResult(_readResult.Buffer, _readResult.IsCanceled, isCompleted: true);
+            }
+
+            if (_readResult.IsCompleted)
+            {
+                TryStop();
+            }
+
+            return _readResult;
+        }
+
+        public override bool TryRead(out ReadResult readResult)
+        {
+            if (_completed)
```

Comment:
I'll consolidate a lot of these checks between ReadAsync and TryRead into helpers.